qualtran/cirq_interop/_cirq_interop.py

```diff
@@ -111,67 +167,85 @@ def cirq_optree_to_cbloq(
+
+    assert signature is not None and cirq_quregs is not None
+
     bb, initial_soqs = BloqBuilder.from_signature(signature, add_registers_allowed=False)
 
     # Magic to make sure signature of the CompositeBloq matches `Signature`.
-    qvars = {}
+    qubit_to_qvar = {}
     for reg in signature.lefts():
+        assert reg.name in cirq_quregs
         soqs = initial_soqs[reg.name]
+        if isinstance(soqs, Soquet):
+            soqs = np.array([soqs])
         if reg.bitsize > 1:
-            # Need to split all soquets here.
-            if isinstance(soqs, Soquet):
-                qvars[reg.name] = bb.split(soqs)
-            else:
-                qvars[reg.name] = np.concatenate([bb.split(soq) for soq in soqs.reshape(-1)])
-        else:
-            if isinstance(soqs, Soquet):
-                qvars[reg.name] = [soqs]
-            else:
-                qvars[reg.name] = soqs.reshape(-1)
-
-    qubit_to_qvar = dict(zip(all_qubits, np.concatenate([*qvars.values()])))
+            soqs = np.array([bb.split(soq) for soq in soqs.flatten()])
+        soqs = soqs.reshape(reg.shape + (reg.bitsize,))
+        assert cirq_quregs[reg.name].shape == soqs.shape
+        qubit_to_qvar |= zip(cirq_quregs[reg.name].flatten(), soqs.flatten())
 
+    allocated_qubits = set()
     for op in circuit.all_operations():
         if op.gate is None:
             raise ValueError(f"Only gate operations are supported, not {op}.")
 
         bloq = CirqGateAsBloq(op.gate)
-        qvars_for_op = np.array([qubit_to_qvar[qubit] for qubit in op.qubits])
-        qvars_for_op_out = bb.add(bloq, qubits=qvars_for_op)
-        qubit_to_qvar |= zip(op.qubits, qvars_for_op_out)
+        for q in op.qubits:
+            if q not in qubit_to_qvar:
+                qubit_to_qvar[q] = bb.add(Allocate(1))
+                allocated_qubits.add(q)
+
+        qvars_in = [qubit_to_qvar[qubit] for qubit in op.qubits]
+        qvars_out = bb.add_t(bloq, **split_qvars(qvars_in, bloq.signature))
+        qubit_to_qvar |= zip(
+            op.qubits, itertools.chain.from_iterable([arr.flatten() for arr in qvars_out])
+        )
 
-    qvar_vals_out = np.array([qubit_to_qvar[qubit] for qubit in all_qubits])
+    for q in allocated_qubits:
+        bb.add(Free(1), free=qubit_to_qvar[q])
 
+    qvars = [*qubit_to_qvar.values()]
     final_soqs = {}
     idx = 0
     for reg in signature.rights():
         name = reg.name
-        soqs = qvar_vals_out[idx : idx + len(qvars[name])]
-        idx = idx + len(qvars[name])
+        assert name in cirq_quregs
+        soqs = qvars[idx : idx + np.product(cirq_quregs[name].shape)]
+        idx = idx + np.product(cirq_quregs[name].shape)
         if reg.bitsize > 1:
             # Need to combine the soquets here.
             if len(soqs) == reg.bitsize:
-                final_soqs[name] = bb.join(soqs)
+                final_soqs[name] = bb.join(np.array(soqs))
```

Comment:
change qvars to a numpy array during construction and then these castings are unnecessary?